File: backend/screen_time_parser.py
"""Turn raw OCR text into a duration in minutes.

Handles notations used by iOS Screen Time and Android Digital Wellbeing
(Samsung, Xiaomi/MI, OnePlus, Realme, Pixel, Nothing, Motorola), as well as
common OCR misreads (e.g. 'n' for 'h', 'rn' for 'm', '04:18', '4 h 18 m').
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import List, Optional

from config import config

# ── Unit vocabulary & Fuzzy OCR variants ─────────────────────────────
# 'h' misreads: h, H, n, N, A, R, k, b, hr, hrs, hour, hours, గం, గంట, గంటలు
_H = r"(?:h|H|n|N|A|R|k|b|hr|hrs|hour|hours|గం|గంట|గంటల)"
# 'm' misreads: m, M, rn, nn, min, mins, minute, minutes, ని, నిమి, నిమిషం, నిమిషాలు
_M = r"(?:m|M|rn|nn|min|mins|minute|minutes|ని|నిమి|నిమిషం|నిమిషాలు)"

# Words that indicate the number belongs to *today's total*
_TODAY_HINTS = (
    "today", "todays", "today's", "daily average", "screen time",
    "screentime", "digital wellbeing", "wellbeing", "ఈరోజు", "నేడు",
    "daily", "average", "total", "usage",
)

# Words that mean the nearby number is NOT today's total
_NEGATIVE_HINTS = ("yesterday", "last week", "weekly", "goal", "limit", "target")

# Digit translation map for common OCR substitutions
_DIGIT_FIXES = str.maketrans({
    "O": "0", "o": "0", "Q": "0", "D": "0",
    "l": "1", "I": "1", "i": "1", "|": "1", "!": "1",
    "Z": "2", "z": "2",
    "S": "5", "s": "5",
    "B": "8",
    "g": "9", "q": "9",
})


@dataclass
class ParsedTime:
    minutes: int
    raw: str                 # exact matched string
    pattern: str             # pattern type
    position: int            # character position
    proximity_bonus: float = 0.0

    @property
    def strength(self) -> float:
        base = {
            "hero_format": 1.0,     # "4h 18m" / "4h 18min" — hero total
            "fuzzy_hm": 0.94,       # "4n 18rn" — OCR misread of 4h 18m
            "colon_format": 0.82,   # "4:18" / "04:18"
            "hours_only": 0.72,     # "5h"
            "minutes_only": 0.48,   # "45m"
        }.get(self.pattern, 0.5)
        return min(1.0, base + self.proximity_bonus)


def _normalize(text: str) -> str:
    text = text.replace("\u00a0", " ")
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def _fix_digits(token: str) -> str:
    return token.translate(_DIGIT_FIXES)


def _proximity(text_low: str, position: int) -> float:
    window = text_low[max(0, position - 80): position + 80]
    bonus = 0.0
    if any(h in window for h in _TODAY_HINTS):
        bonus += 0.25
    if any(n in window for n in _NEGATIVE_HINTS):
        bonus -= 0.50
    return bonus


def find_all(text: str) -> List[ParsedTime]:
    """Extract every plausible screen time duration from OCR text."""
    text = _normalize(text)
    low = text.lower()
    found: List[ParsedTime] = []

    def add(minutes: int, raw: str, pattern: str, pos: int) -> None:
        if not (config.MIN_PLAUSIBLE_MINUTES <= minutes <= config.MAX_PLAUSIBLE_MINUTES):
            return
        found.append(
            ParsedTime(
                minutes=minutes,
                raw=raw.strip(),
                pattern=pattern,
                position=pos,
                proximity_bonus=_proximity(low, pos),
            )
        )

    # 1) Exact "4h 18m", "4 h 18 min", "4 hrs 18 mins", "4గ 18ని"
    for m in re.finditer(rf"(\d{{1,2}})\s*{_H}\s*(\d{{1,2}})\s*{_M}?", low, re.I):
        try:
            h = int(_fix_digits(m.group(1)))
            mins = int(_fix_digits(m.group(2)))
            if mins < 60:
                add(h * 60 + mins, m.group(0), "hero_format", m.start())
        except ValueError:
            pass

    # 2) Joined / Fuzzy "4h18m", "04h18m", "4n18rn", "4h 18"
    for m in re.finditer(r"(?<!\d)(\d{1,2})\s*([hn]|hr|hrs)\s*(\d{1,2})\s*(m|min|mins|rn)?(?!\w)", low, re.I):
        try:
            h = int(_fix_digits(m.group(1)))
            mins = int(_fix_digits(m.group(3)))
            if mins < 60:
                add(h * 60 + mins, m.group(0), "fuzzy_hm", m.start())
        except ValueError:
            pass

    # 3) Clock / Duration format "4:18", "04:18", "4 : 18", "4.18", "4-18"
    for m in re.finditer(r"(?<!\d)(\d{1,2})\s*[:;.\-]\s*([0-5]\d)(?!\d)", low):
        try:
            h = int(m.group(1))
            mins = int(m.group(2))
            if h <= 23:
                add(h * 60 + mins, m.group(0), "colon_format", m.start())
        except ValueError:
            pass

    # 4) Hours only: "5h", "6 hr", "7 hours", "5 గంటలు"
    for m in re.finditer(rf"(\d{{1,2}})\s*{_H}(?!\s*\d)", low, re.I):
        try:
            add(int(_fix_digits(m.group(1))) * 60, m.group(0), "hours_only", m.start())
        except ValueError:
            pass

    # 5) Minutes only: "45m", "45 min", "45 నిమిషాలు"
    for m in re.finditer(rf"(?<!\d)(?<![hHnN]\s)(\d{{1,3}})\s*{_M}(?![a-z])", low, re.I):
        try:
            mins = int(_fix_digits(m.group(1)))
            if mins < 600:
                add(mins, m.group(0), "minutes_only", m.start())
        except ValueError:
            pass

    return found


def select_best(candidates: List[ParsedTime]) -> Optional[ParsedTime]:
    if not candidates:
        return None
    # Rank by strength (pattern + proximity hint), then pick total minutes
    ranked = sorted(candidates, key=lambda c: (round(c.strength, 3), c.minutes), reverse=True)
    top = ranked[0]
    peers = [c for c in ranked if abs(c.strength - top.strength) < 0.10]
    return max(peers, key=lambda c: c.minutes) if peers else top


def parse(text: str) -> Optional[ParsedTime]:
    return select_best(find_all(text))

File: src/services/screenTimeParser.ts
/**
 * High-performance Screen Time parser for browser WASM & frontend.
 * Matches iOS Screen Time and Android Digital Wellbeing (Samsung, Xiaomi,
 * OnePlus, Realme, Pixel, Nothing, Moto), including fuzzy OCR substitutions.
 */

export interface ParsedTime {
  minutes: number;
  raw: string;
  pattern: 'hero_format' | 'fuzzy_hm' | 'colon_format' | 'hours_only' | 'minutes_only';
  position: number;
  proximityBonus: number;
}

const H = '(?:h|H|n|N|A|R|k|b|hr|hrs|hour|hours|గం|గంట|గంటల)';
const M = '(?:m|M|rn|nn|min|mins|minute|minutes|ని|నిమి|నిమిషం|నిమిషాలు)';

const TODAY_HINTS = [
  'today', 'todays', "today's", 'daily average', 'screen time',
  'screentime', 'digital wellbeing', 'wellbeing', 'ఈరోజు', 'నేడు',
  'daily', 'average', 'total', 'usage',
];
const NEGATIVE_HINTS = ['yesterday', 'last week', 'weekly', 'goal', 'limit', 'target'];

const BASE_STRENGTH: Record<ParsedTime['pattern'], number> = {
  hero_format: 1.0,     // "4h 18m" / "4h 18min"
  fuzzy_hm: 0.94,       // "4n 18rn" (OCR misread)
  colon_format: 0.82,   // "4:18"
  hours_only: 0.70,     // "5h"
  minutes_only: 0.48,   // "45m"
};

export const MIN_PLAUSIBLE_MINUTES = 1;
export const MAX_PLAUSIBLE_MINUTES = 24 * 60;

function fixDigits(token: string): string {
  return token
    .replace(/[OoQD]/g, '0')
    .replace(/[lIi|!]/g, '1')
    .replace(/[Zz]/g, '2')
    .replace(/[Ss]/g, '5')
    .replace(/B/g, '8')
    .replace(/[gq]/g, '9');
}

function normalize(text: string): string {
  return text.replace(/\u00a0/g, ' ').replace(/\s+/g, ' ').trim();
}

function proximity(lowText: string, position: number): number {
  const window = lowText.slice(Math.max(0, position - 80), position + 80);
  let bonus = 0;
  if (TODAY_HINTS.some((h) => window.includes(h))) bonus += 0.25;
  if (NEGATIVE_HINTS.some((n) => window.includes(n))) bonus -= 0.50;
  return bonus;
}

export function strengthOf(p: ParsedTime): number {
  return Math.min(1, BASE_STRENGTH[p.pattern] + p.proximityBonus);
}

export function findAll(text: string): ParsedTime[] {
  const clean = normalize(text);
  const low = clean.toLowerCase();
  const out: ParsedTime[] = [];

  const add = (minutes: number, raw: string, pattern: ParsedTime['pattern'], position: number) => {
    if (minutes < MIN_PLAUSIBLE_MINUTES || minutes > MAX_PLAUSIBLE_MINUTES) return;
    out.push({ minutes, raw: raw.trim(), pattern, position, proximityBonus: proximity(low, position) });
  };

  // 1) "4h 18m" / "3 hrs 20 mins"
  for (const m of low.matchAll(new RegExp(`(\\d{1,2})\\s*${H}\\s*(\\d{1,2})\\s*${M}?`, 'gi'))) {
    const h = parseInt(fixDigits(m[1]), 10);
    const mins = parseInt(fixDigits(m[2]), 10);
    if (!isNaN(h) && !isNaN(mins) && mins < 60) {
      add(h * 60 + mins, m[0], 'hero_format', m.index ?? 0);
    }
  }

  // 2) Joined/Fuzzy "4h18m", "4n18rn", "4h 18"
  for (const m of low.matchAll(/(?<!\d)(\d{1,2})\s*([hn]|hr|hrs)\s*(\d{1,2})\s*(m|min|mins|rn)?(?!\w)/gi)) {
    const h = parseInt(fixDigits(m[1]), 10);
    const mins = parseInt(fixDigits(m[3]), 10);
    if (!isNaN(h) && !isNaN(mins) && mins < 60) {
      add(h * 60 + mins, m[0], 'fuzzy_hm', m.index ?? 0);
    }
  }

  // 3) "4:18" / "04:18" / "4.18" / "4-18"
  for (const m of low.matchAll(/(?<!\d)(\d{1,2})\s*[:;.\-]\s*([0-5]\d)(?!\d)/g)) {
    const h = parseInt(m[1], 10);
    const mins = parseInt(m[2], 10);
    if (!isNaN(h) && !isNaN(mins) && h <= 23) {
      add(h * 60 + mins, m[0], 'colon_format', m.index ?? 0);
    }
  }

  // 4) Hours only: "5h", "6 hr"
  for (const m of low.matchAll(new RegExp(`(\\d{1,2})\\s*${H}(?!\\s*\\d)`, 'gi'))) {
    const h = parseInt(fixDigits(m[1]), 10);
    if (!isNaN(h)) add(h * 60, m[0], 'hours_only', m.index ?? 0);
  }

  // 5) Minutes only: "45m", "45 min"
  for (const m of low.matchAll(new RegExp(`(?<!\d)(?<![hHnN]\\s)(\\d{1,3})\\s*${M}(?![a-z])`, 'gi'))) {
    const mins = parseInt(fixDigits(m[1]), 10);
    if (!isNaN(mins) && mins < 600) add(mins, m[0], 'minutes_only', m.index ?? 0);
  }

  return out;
}

export function selectBest(candidates: ParsedTime[]): ParsedTime | null {
  if (!candidates.length) return null;
  const ranked = [...candidates].sort((a, b) => {
    const d = strengthOf(b) - strengthOf(a);
    return Math.abs(d) > 1e-6 ? d : b.minutes - a.minutes;
  });
  const top = ranked[0];
  const peers = ranked.filter((c) => Math.abs(strengthOf(c) - strengthOf(top)) < 0.10);
  return peers.reduce((best, c) => (c.minutes > best.minutes ? c : best), peers[0] ?? top);
}

export function parse(text: string): ParsedTime | null {
  return selectBest(findAll(text));
}

export function countContext(text: string): number {
  const low = text.toLowerCase();
  return TODAY_HINTS.reduce((n, kw) => (low.includes(kw) ? n + 1 : n), 0);
}

export function formatDuration(totalMinutes: number): string {
  const t = Math.max(0, Math.round(totalMinutes));
  const h = Math.floor(t / 60);
  const m = t % 60;
  if (h && m) return `${h}h ${m}m`;
  if (h) return `${h}h`;
  return `${m}m`;
}
